cumulus/pallets/xcmp-queue/src/weights_ext.rs

```diff
@@ -58,36 +59,49 @@ pub trait WeightInfoExt: WeightInfo {
 				.saturating_sub(Self::enqueue_n_bytes_xcmp_message(0))
 		};
 
-		pages_overhead.saturating_add(messages_overhead).saturating_add(bytes_overhead)
+		// If the messages are not added to the beginning of the first page, the page will be
+		// decoded and re-encoded once. Let's account for this.
+		let pos_overhead = {
+			Self::enqueue_empty_xcmp_message_at(first_page_pos)
+				.saturating_sub(Self::enqueue_empty_xcmp_message_at(0))
+		};
+
+		pages_overhead
+			.saturating_add(messages_overhead)
+			.saturating_add(bytes_overhead)
+			.saturating_add(pos_overhead)
+	}
+
+	#[cfg(feature = "std")]
+	fn check_accuracy<MaxMessageLen: bounded_collections::Get<u32>>(err_margin: u8) {
+		assert!(err_margin < 100);
+		let err_margin = err_margin as u64;
+
+		let estimated_weight =
+			Self::uncached_enqueue_xcmp_messages().saturating_add(Self::enqueue_xcmp_messages(
+				get_average_page_pos(MaxMessageLen::get()),
+				&BatchFootprint { msgs_count: 1000, size_in_bytes: 3000, new_pages_count: 0 },
+			));
+		let actual_weight = Self::enqueue_1000_small_xcmp_messages();
+
+		// Check that the ref_time diff is less than {err_margin}%
+		let diff_ref_time = estimated_weight.ref_time().abs_diff(actual_weight.ref_time());
+		assert!(diff_ref_time < estimated_weight.ref_time() * err_margin / 100);
+		assert!(diff_ref_time < actual_weight.ref_time() * err_margin / 100);
 	}
 }
 
 impl<T: WeightInfo> WeightInfoExt for T {}
 
-#[cfg(feature = "std")]
-pub fn check_weight_info_ext_accuracy<T: WeightInfoExt>(err_margin: u8) {
-	assert!(err_margin < 100);
-	let err_margin = err_margin as u64;
-
-	let estimated_weight =
-		T::uncached_enqueue_xcmp_messages().saturating_add(T::enqueue_xcmp_messages(1, 1000, 3000));
-	let actual_weight = T::enqueue_1000_small_xcmp_messages();
-
-	// Check that the ref_time diff is less than {err_margin}%
-	let diff_ref_time = estimated_weight.ref_time().abs_diff(actual_weight.ref_time());
-	assert!(diff_ref_time < estimated_weight.ref_time() * err_margin / 100);
-	assert!(diff_ref_time < actual_weight.ref_time() * err_margin / 100);
-
-	// The proof sizes should be the same
-	assert_eq!(estimated_weight.proof_size(), actual_weight.proof_size());
-}
-
 #[cfg(test)]
 mod tests {
 	use super::*;
+	use bounded_collections::ConstU32;
 
 	#[test]
-	fn weight_info_ext_accuracy_is_high() {
-		check_weight_info_ext_accuracy::<()>(5);
+	fn check_weight_info_ext_accuracy() {
+		// The `MaxMessageLen` was manually copied from `asset-hub-westend-runtime`.
```

Comment:
Can you put a permalink to the line of code that it was from please?

Reply:
Done.